src/sources.rs

```diff
@@ -0,0 +1,346 @@
+ */
+pub async fn pull_exchange_rate(
+    stats: prometheus::Stats,
+    source: Source,
+    rate_history_mutex: Arc<Mutex<VecDeque<BigRational>>>,
+    pull_interval: u32,
+    max_rates_saved: usize,
+    mut database_conn: Option<mysql::PooledConn>,
+) -> anyhow::Result<()> {
+    let client = reqwest::Client::new();
+    let source_label = match source {
+        Source::Bitfinex => config::BITFINEX_LABEL,
+        Source::LiveCoinWatch(_) => config::LIVECOINWATCH_LABEL,
+        Source::CoinMarketCap(_) => config::COINMARKETCAP_LABEL,
+        Source::CoinGecko => config::COINGECKO_LABEL,
+        Source::Test(_, ref id) => &id,
+    };
+
+    let mut interval = interval(Duration::from_secs(pull_interval.into()));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+    loop {
+        interval.tick().await;
+        log::debug!("{}: Polling for exchange rate", source_label);
+
+        let raw_rate = match request_with_backoff(
+            || request_matcher(client.clone(), source.clone()),
+            |timeout: u64| {
+                log::warn!(
+                    "{}: Request not successful. Waiting for {} seconds until trying again",
+                    source_label,
+                    timeout
+                );
+                stats.increment_read_attempts(source_label);
+            },
+            INITIAL_RETRY_INTERVAL,
+            MAX_RETRIES,
+        )
+        .await
+        {
+            Some(i) => i,
+            None => {
+                log::error!("{}: Request failed. Retries exhausted", source_label);
+                stats.increment_read_attempts(source_label);
+                continue;
+            }
+        };
+        stats.reset_read_attempts(source_label);
+
+        if let Some(ref mut conn) = database_conn {
+            if let Err(e) = crate::database::write_read_rate(conn, raw_rate, source_label) {
+                stats.increment_failed_database_updates();
+                log::error!(
+                    "{}: Unable to INSERT new reading: {}, due to: {}",
+                    source_label,
+                    raw_rate,
+                    e
+                )
+            };
+        }
+        stats.update_read_rate(raw_rate, &source_label);
+
+        let rate = match BigRational::from_float(raw_rate) {
+            Some(r) => r.recip(), // Get the inverse value, to change units from EUR/CCD to CCD/EUR
+            None => {
+                log::error!("{}: Unable to convert rate to rational: {}", source_label, raw_rate);
+                continue;
+            }
+        };
+        log::info!("{}: New exchange rate polled: {}/{}", source_label, rate.numer(), rate.denom());
+        {
+            let mut rates = rate_history_mutex.lock().unwrap();
+            rates.push_back(rate);
+            if rates.len() > max_rates_saved {
+                rates.pop_front();
+            }
+        }; // drop lock
```

Comment:
I think the semicolon there is redundant.

Reply:
It is. I would've thought clippy would complain :shrug: